cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,948 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <algorithm>
+#include <execution>
+#include <optional>
+#include <raft/core/cuda_support.hpp>
+#include <raft/core/detail/copy.hpp>
+#include <raft/core/device_container_policy.hpp>
+#include <raft/core/error.hpp>
+#include <raft/core/host_container_policy.hpp>
+#include <raft/core/host_device_accessor.hpp>
+#include <raft/core/logger.hpp>
+#include <raft/core/managed_container_policy.hpp>
+#include <raft/core/mdarray.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/pinned_container_policy.hpp>
+#include <raft/core/stream_view.hpp>
+#include <raft/util/variant_utils.hpp>
+#include <type_traits>
+#include <utility>
+#include <variant>
+#ifndef RAFT_DISABLE_CUDA
+#include <raft/core/copy.cuh>
+#include <raft/util/cudart_utils.hpp>
+#include <thrust/device_ptr.h>
+#else
+#include <raft/core/copy.hpp>
+#endif
+
+namespace raft {
+
+/**
+ * @defgroup mdbuffer_apis multi-dimensional maybe-owning type
+ * @{
+ */
+
+/**
+ * @brief Retrieve a canonical index associated with a given memory type.
+ *
+ * For variants based on memory type, this index can be used to help keep a
+ * consistent ordering of the memory types in the variant.
+ */
+inline auto constexpr variant_index_from_memory_type(raft::memory_type mem_type)
+{
+  return static_cast<std::underlying_type_t<raft::memory_type>>(mem_type);
+}
+
+/**
+ * @brief Retrieve a type from a variant based on a given memory type.
+ */
+template <raft::memory_type MemType, typename Variant>
+using alternate_from_mem_type =
+  std::variant_alternative_t<variant_index_from_memory_type(MemType) % std::variant_size_v<Variant>,
+                             Variant>;
+
+/**
+ * @brief A variant of container policies for each memory type which can be
+ * used to build the default container policy for a buffer.
+ */
+template <typename T>
+using default_container_policy_variant = std::variant<host_vector_policy<T>,
+                                                      device_uvector_policy<T>,
+                                                      managed_uvector_policy<T>,
+                                                      pinned_vector_policy<T>>;
```

Comment:
I feel a little bit uncomfortable that this variant ordering seems to be tied to the `variant_index_from_memory_type` and the `raft::memory_type` declared in `core/memory_type.hpp`, yet there's no static (compile time) enforcing of that.
Have you considered somehow addressing this? e.g. some kind of `static_assert` just below the declaration that would combine `variant_index_from_memory_type` and `std::get (std::variant)` for every alternative to ensure the connection and catch any changes to one of these if they ever happen?

Reply:
Agree entirely! Take a look at the change I just pushed and see what you think. I believe it should ensure that we get the correct policy even if the enum is reordered.

Reply:
This should work, thanks!